RFC for profile directed typing using Monkeytype tool in TorchScript (#23)
pytorch/rfcs: RFC-0012-profile-directed-typing.md

## Overview

> Missing/incorrect type annotations is a common reason for failed model scripting. For a long time, the process of mitigating missing/incorrect type annotations has been through trial and error, ie. by repeatedly hitting the wall of various type-checking errors generated from torch.jit.script and fixing them one by one. This is both inefficient and frustrating.

Reviewer: I would prefer "A common reason for failed model scripting is missing or incorrect type annotations", because then the first part of the sentence is sort of a "build up" to the solution we're proposing

* PyAnnotate is owned by DropBox, while MonkeyType is maintained by Facebook/Instagram, therefore we are likely going to get better support.

## Bidirectional Type Inference Algorithm

This is another approach of solving the general problem of missing/incorrect type annotations. Basic idea is to devise an algorithm that infers type of an argument based on how it is used in function body. This approach carries a lot higher complexity and may not be a good fit for TorchScript language specifically.

Stand-alone Tool with Human Intervention to Permanently Annotate Source Code

Another option we thought about is to create a stand-alone tool that is based on MonkeyType to help users permanently annotate their source code, instead of only generating one-time type annotations on the fly as we proposed in this approach.

The additional benefit is that stand-alone tool could optionally ask for user’s opinion on whether annotated types match their expectation and make revision based on user input.

However, using a stand-alone tool has a few down-sides:

> * MonkeyType is unaware of TorchScript limitation of expressing nn.Module types and would blindly apply nn.Module and its subclasses as type annotation. Though legal in Python, they are not accepted in TorchScript. We need to configure MonkeyType to ensure it never annotates module types.

Reviewer: ```suggestion
* MonkeyType is unaware of the TorchScript limitation of expressing nn.Module types and would blindly apply nn.Module and its subclasses as type annotations. Though legal in Python, they are not accepted in TorchScript. We need to configure MonkeyType to ensure it never annotates module types.
```

> * TorchScript has sophisticated type inference for return values of functions/methods, empirically it is more than enough to infer all return types of functions. Having MonkeyType annotate return types is more likely to cause type conflicts rather than helping with compilation.

Reviewer: ```suggestion
* TorchScript has a sophisticated type inference for return values that is more than enough to infer all return types of functions. Having MonkeyType annotate return types is more likely to cause type conflicts than helping with compilation.
```

> * User code may reference third-party libraries that user installed in system-wide locations. If we have a stand-alone tool that profiles types and applies derived type annotations to them, then user environment is polluted. If type annotations are not applied to them, then TorchScript would fail.

Reviewer: ```suggestion
* User code may reference third-party libraries that user has installed in system-wide locations. If we have a stand-alone tool that profiles types and applies derived type annotations, then the user environment would be polluted. On the other hand, if the type annotations were not applied, then TorchScript would fail.
```

> * Using a stand-alone tool is a less stream-lined experience compared to type inference embedded in torch.jit.script calls because users need to create another program that drives compilation target function.

Reviewer: *streamlined

> * Intuitively I feel creating such stand-alone tool belongs and quip it with human intervention possibility is more like an enhancement to MonkeyType rather than being a part of PyTorch.

Reviewer: opinion, please delete
